### DISCUSSION TITLE
Using an external eye-tracker with jsPsych?

### DISCUSSION THREAD

**Discussion by prashantig25** [jspsych/jsPsych#2226] [Q&A]
Hello All,
I have a behavioural task that runs using jsPsych. However, I would like to use the same code to conduct a lab study but using an eye-tracking set-up. I would like to use the lab's eye-tracker and not the web-cam. Is there a way to do that? How can I connect an external code so that it works with the jsPsych code? Any help is appreciated.
Thank you!!
Sincerely,
Prashanti

**Comment by jodeleeuw** [comment]
Hi @prashantig25,

This will depend on the eye tracker you are using. It'll also require adding some custom support in all likelihood. 

We are working on adding Lab Streaming Layer support via a python based webserver for locally hosted jsPsych experiments. If your eye tracker supports LSL then we could talk about testing our LSL server with your eye tracker.

**Reply by prashantig25** [reply]
Hi @jodeleeuw,

Thank you for your response. I am planning to use the SMI eye-tracker, which I believe supports LSL. Please let me know if we can test your LSL server with this eye-tracker. 

Thank you.
Sincerely,
Prashanti

**Reply by selinbekir** [reply]
Hi @jodeleeuw I have the same question! We want to keep using jsPsych in our in-lab studies as well (as well as using cognition.run as the platform in which we run studies). We want to use an eye-tracker but I wasn't sure how to incorporate that with jsPsych and cognition.run. Looking forward for the updates on this!
Thanks,
Selin

**Reply by jodeleeuw** [reply]
Hi @selinbekir. What eye tracking system are you using in your lab?

**Reply by selinbekir** [reply]
Hi again @jodeleeuw we will be using Eye Link (https://www.sr-research.com/eyelink-1000-plus/) in this experiment. And i am using cognition.run as my platform to run studies. I am just not sure how to set them up together, or if that is possible.

Thank you!
Selin

**Reply by jodeleeuw** [reply]
I don't think it will be possible to use cognition.run and use Eye Link at the same time. To do that you would need to somehow communicate between the eye tracker and cognition.run directly, and I can't see any way to reasonably do that. 

The solution that we use for EEG in my lab is to host the jsPsych experiment locally using a python-based webserver (like Flask). JavaScript can then trigger events on the webserver, which can then communicate with the EEG equipment using python code. It looks like such a thing should be possible with Eye Link using their `pylink` package, but it would require some (potentially substantial) customization to get up and running.

**Reply by selinbekir** [reply]
Oh okay, good to know, I will try that. Thank you very much for your responses @jodeleeuw it's been very helpful!

**Reply by csavasegal** [reply]
I was wondering if anyone on this thread can point me towards sample code for integrating pylink with Flask? I currently have a Flask app working and linking to my hosted jspsych experiment, but am stuck on the next steps, including what pylink function to call post-calibration if I'm hoping to send a trigger on each trial (presumably in the `on_load`)

I'm also using the eyelink eyetracker mentioned above. Thanks so much!

**Reply by emid** [reply]
@jodeleeuw  We are planning to integrate jspsych experiments with LSL. How can we test that python based webserver for locally hosted jsPsych experiments?

**Reply by derekalbert** [reply]
@jodeleeuw, did you ever get the LSL integration working? I currently have a similar use case for it. Thanks for all your wonderful work.

**Comment by fredcallaway** [comment]
I'm also trying to get this to work (with EyeLink). @csavasegal did you make progress?

**Reply by csavasegal** [reply]
Hi @fredcallaway. Yes, I did! 

Essentially I wrapped my main JsPsych code in a Flask app that launched the eyetracker and then launched the experiment. I sent triggers from my experiment back to the eyetracker so that I knew when certain trials were being launched/ended (put these within the "on load"/"on finish" blocks)

Let me know if you have follow-up questions! Happy to share code snippets, as well. 


P.S. Depending on your usage, I've since learned that most people calibrate with psychopy. I was unable to do that given my Flask app, so I wound up calibrating using pygame. This was a stressor for me so I thought I'd pass that along! :)

**Reply by fredcallaway** [reply]
That's great to hear!

Any code you'd be willing to share would be tremendously helpful. I have a flask app already set up (using psiturk). 

For calibration, I was thinking I would calibrate in javascript so that I could easily check for drift and recalibrate within a session. But this would require implementing an eyelink-compatible graphics  in js (might not be _so_ bad; [this psychopy implementation](https://github.com/colinquirk/templateexperiments/blob/master/eyelinker/PsychoPyCustomDisplay.py) is only ~200 lines). I'm curious how you handled this though. Did you just calibrate once and get good results? Or did you have participants tab out to recalibrate in pygame occasionally?

EDIT: Sorry one more question. Can you say anything more about why pygame was difficult?

**Reply by csavasegal** [reply]
I've just been splitting into runs, and calibrating at the start of each run, which has worked fine on my end!

**Reply by atakaragoz** [reply]
Hello,

Sorry to resurface a several month old thread. I've been trying to use a custom flask app for EEG as well as eyetracking with jsPsych. I've had success using pySerial to send triggers using the com port in the EEG case. Was wondering if y'all had any code you'd be willing to share for interfacing with the eye link?

On another note I ran some timing checks with a photodiode on the EEG and found that the latency between jsPsych flipping the screen and the trigger being read by the EEG rig seems to have a ~16ms jitter (this might be due to monitor refresh, although the monitor is supposedly 120hz). Was wondering if timing sensitivity ended up being an issue with either of you?

Would tremendously appreciate any help! 

Best,
Ata

**Reply by martonaronvarga** [reply]
Hi,

I would like to follow up on this as well, I am currently running an online experiment using Express as backend and jsPsych as frontend, with good latencies. However, I am thinking of incorporating EEG into the paradigm, but I am worried about hosting and communicating with the hardware using a server would introduces lags to the system. In my paradigm, timings matter a lot. What do you recommend to avoid on-line communication and switch to local hardware support without introducing lags?

**Reply by fredcallaway** [reply]
Not sure if this is useful, but I put pylink trigger into my Flask app and was getting substantial latencies, in the 100ms range. The latency is highly variable, so it's tricky to account for it. Basically, if the relative timing of on-screen events and your external data source is important, I don't advise this approach. It's possible that this was specific to my setup, but my current advice is to bite the bullet and reimplement your task in pschopy or pygame :\

**Comment by DominiqueMakowski** [comment]
Just checking for any update made with external devices that could be synced with LSL?
We're considering getting a Tobii eye-tracking bar, and I was  wondering about the  possibility of using that instead of the default webcam.
